### PR TITLE
Hide passphrase from host

### PR DESCRIPTION
**Pull request by grdddj**
Fixes https://github.com/trezor/trezor-firmware/issues/2598:
- adds the optional possibility of hiding the passphrase coming from host

Allows for changing this behavior by `trezorctl set passphrase hide {on,off}`.
It is currently hidden behind `safety checks`.
It is being shown in `Features` as `hide_passphrase_from_host: {True,False}`.

Trezorlib step-by-step flow I used when testing this:
- $ `trezorctl device wipe && trezorctl device load -s && trezorctl set pin`
- $ `trezorctl get-features | grep passphrase_protection` 
  - `passphrase_protection: False`
- $ `trezorctl set passphrase on` 
- $ `trezorctl get-features | grep passphrase_protection` 
  - `passphrase_protection: True`
- $ `PASSPHRASE=abc trezorctl -P btc get-address -n "84'/0'/0'/1/0"`
  - shows `Next screen will show the passphrase`
- $ `trezorctl get-features | grep hide_passphrase` 
  - `hide_passphrase_from_host: False`
- $ `trezorctl set passphrase hide on` 
  - `Error: ProcessError: Safety checks are strict` 
- $ `trezorctl set safety-checks prompt` 
  - `Settings applied`   
- $ `trezorctl set passphrase hide on` 
  - `Settings applied`
- $ `trezorctl get-features | grep hide_passphrase` 
  - `hide_passphrase_from_host: True`
- $ `PASSPHRASE=abc trezorctl -P btc get-address -n "84'/0'/0'/1/0"`
  - shows `Passphrase provided by host will be used but will not be displayed due to the device settings.`

There are currently no changelog entries (and `[no changelog]` in both commits), as I understand it we do not want to reveal it to the public if not necessary.

### CODE REVIEW

File: core/src/apps/management/apply_settings.py [hunk 1]
```diff
@@ -269,3 +277,18 @@ async def _require_confirm_experimental_features(
...
+
+async def _require_confirm_hide_passphrase_from_host(
+    ctx: GenericContext, enable: bool
+) -> None:
+    if enable:
+        await confirm_action(
+            ctx,
+            "set_hide_passphrase_from_host",
+            "Hide passphrase",
+            "Only for development and beta testing!",
```
Comment (matejcik) [inline core/src/apps/management/apply_settings.py:290]: it's not really. I think we can completely remove this line?

Comment (grdddj) [reply core/src/apps/management/apply_settings.py]: Removed in [3759d83](https://github.com/trezor/trezor-firmware/pull/2829/commits/3759d83cf432023988b410f13fbcaf326dc875d0)

File: core/src/apps/management/apply_settings.py [hunk 2]
```diff
@@ -117,6 +119,12 @@ async def apply_settings(ctx: Context, msg: ApplySettings) -> Success:
         await _require_confirm_experimental_features(ctx, experimental_features)
         storage_device.set_experimental_features(experimental_features)
 
+    if hide_passphrase_from_host is not None:
+        if not storage_device.get_experimental_features():
```
[the remark this answers is not in the document]
Comment (grdddj) [reply core/src/apps/management/apply_settings.py]: Changed in [3759d83](https://github.com/trezor/trezor-firmware/pull/2829/commits/3759d83cf432023988b410f13fbcaf326dc875d0)

**Comment by grdddj** [comment]
Turned out I was mistaking `experimental features` and `safety checks` - updated the commands above to `safety checks`.

Also screenshots of the two new screens:

image image



Do we want to have UI tests for this?

**Comment by matejcik** [comment]
> Do we want to have UI tests for this?

it would be nice, yes.

**Comment by grdddj** [comment]
> > Do we want to have UI tests for this?
> 
> it would be nice, yes.

Added in [b94c099](https://github.com/trezor/trezor-firmware/pull/2829/commits/b94c0995ee66597fa227819751ce10143e7efe83)

File: tests/device_tests/test_session_id_and_passphrase.py [hunk 3]
```diff
@@ -375,6 +376,40 @@ def call(passphrase: str, expected_result: bool):
...
+        device.apply_settings(client, hide_passphrase_from_host=True)
+
+    device.apply_settings(client, safety_checks=SafetyCheckLevel.PromptTemporarily)
+
+    # Turning it on
+    device.apply_settings(client, hide_passphrase_from_host=True)
+
+    passphrase = "abc"
+
+    response = client.call_raw(XPUB_REQUEST)
```
Comment (matejcik) [inline tests/device_tests/test_session_id_and_passphrase.py:392]: would be nice to also assert on the text on Trezor screen, but I'm not sure how difficult that is to add here.
please try but if it's more than "easy", let's not stop the PR on it

File: tests/device_tests/test_session_id_and_passphrase.py [hunk 4]
```diff
@@ -375,6 +376,40 @@ def call(passphrase: str, expected_result: bool):
     call(passphrase="A" * 49 + "š", expected_result=False)
 
 
+@pytest.mark.setup_client(passphrase=True)
+def test_hide_passphrase_from_host(client: Client):
```
Comment (matejcik) [inline tests/device_tests/test_session_id_and_passphrase.py:380]: pretty sure T1 is going to fail on this. we don't want the feature for T1 so please add `skip_t1`

Comment (grdddj) [reply tests/device_tests/test_session_id_and_passphrase.py:381]: Right, added in [c2cd56a](https://github.com/trezor/trezor-firmware/pull/2829/commits/c2cd56a68a607224e2d392b32d60c2bc7481cb9d)

File: tests/device_tests/test_session_id_and_passphrase.py (hunk 3 above)
Comment (grdddj) [reply tests/device_tests/test_session_id_and_passphrase.py]: It was possible with the use of `input_flows` - the testing code is now much longer, but more thorough, as it also tests the amount of `ButtonRequests` that differs according to `hide` option. Done in [c2cd56a](https://github.com/trezor/trezor-firmware/pull/2829/commits/c2cd56a68a607224e2d392b32d60c2bc7481cb9d)

**Review by matejcik — APPROVED** [review]
perfect :100:

**Comment by bosomt** [comment]
**QA OK**

- works as expected
- only noted that when you reconnect device safety checks are turned off but `trezorctl set passphrase hide on` setting is remembered, i hope its intentional.
- great feature!

**Info:**
 - Suite version: desktop 23.3.0 (8219f5c62d380a3be25dc8cfc51dad10a427b51e)
 - Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) TrezorSuite/23.3.0 Chrome/104.0.5112.124 Electron/20.3.12 Safari/537.36
 - OS: MacIntel
 - Screen: 1440x900
 - Device: model T 2.5.4 Universal (revision 0a955b4ad6d9fb5e7a1dbd3333548ffd8b33399a)
 - Transport: NodeUsbTransport
 - trezorctl 0.13.6 latest master

**Pull request merged**